File: src/ClaudeToolbar.Core/Refresh/RefreshScheduler.cs
using ClaudeToolbar.Core.Time;

namespace ClaudeToolbar.Core.Refresh;

/// <summary>Decides when the next usage fetch is due. Not thread-safe; call from one thread.</summary>
public sealed class RefreshScheduler
{
    public static readonly TimeSpan InitialBackoff = TimeSpan.FromSeconds(15);
    public static readonly TimeSpan MaxBackoff = TimeSpan.FromSeconds(300);
    private static readonly TimeSpan ResetGrace = TimeSpan.FromSeconds(1);

    private readonly IClock _clock;

    public RefreshScheduler(IClock clock, int intervalSeconds = 60)
    {
        _clock = clock;
        IntervalSeconds = intervalSeconds;
        NextDue = clock.UtcNow;
    }

    public int IntervalSeconds { get; set; }

    public DateTimeOffset? NextDue { get; private set; }

    public bool IsPaused => NextDue is null;

    public TimeSpan CurrentBackoff { get; private set; } = TimeSpan.Zero;

    public bool IsDue(DateTimeOffset now) => NextDue is { } due && now >= due;

    public void RequestImmediate() => NextDue = _clock.UtcNow;

    public void Pause() => NextDue = null;

    public void OnSuccess(DateTimeOffset? nextReset)
    {
        CurrentBackoff = TimeSpan.Zero;
        var now = _clock.UtcNow;
        var due = now + TimeSpan.FromSeconds(IntervalSeconds);
        if (nextReset is { } reset)
        {
            var afterReset = reset + ResetGrace;
            if (afterReset > now && afterReset < due) due = afterReset;
        }
        NextDue = due;
    }

    public void OnFailure(TimeSpan? retryAfter)
    {
        CurrentBackoff = CurrentBackoff == TimeSpan.Zero
            ? InitialBackoff
            : TimeSpan.FromTicks(Math.Min((CurrentBackoff * 2).Ticks, MaxBackoff.Ticks));
        var delay = retryAfter is { } ra && ra > CurrentBackoff ? ra : CurrentBackoff;
        NextDue = _clock.UtcNow + delay;
    }
}

File: mac/Sources/ClaudeToolbarCore/Refresh/RefreshScheduler.swift
import Foundation

/// Decides when the next usage fetch is due. Owned and used only by `UsageMonitor`.
public final class RefreshScheduler {
    public static let initialBackoff: TimeInterval = 15
    public static let maxBackoff: TimeInterval = 300
    static let resetGrace: TimeInterval = 1

    private let clock: any ClockSource

    public var intervalSeconds: Int
    public private(set) var nextDue: Date?
    public private(set) var currentBackoff: TimeInterval = 0

    public init(clock: any ClockSource, intervalSeconds: Int = 60) {
        self.clock = clock
        self.intervalSeconds = intervalSeconds
        nextDue = clock.now
    }

    public var isPaused: Bool { nextDue == nil }

    public func isDue(at now: Date) -> Bool {
        guard let due = nextDue else { return false }
        return now >= due
    }

    public func requestImmediate() {
        nextDue = clock.now
    }

    public func pause() {
        nextDue = nil
    }

    public func onSuccess(nextReset: Date?) {
        currentBackoff = 0
        let now = clock.now
        var due = now.addingTimeInterval(TimeInterval(intervalSeconds))
        if let reset = nextReset {
            let afterReset = reset.addingTimeInterval(Self.resetGrace)
            if afterReset > now && afterReset < due { due = afterReset }
        }
        nextDue = due
    }

    public func onFailure(retryAfter: TimeInterval?) {
        currentBackoff = currentBackoff == 0 ? Self.initialBackoff : min(currentBackoff * 2, Self.maxBackoff)
        var delay = currentBackoff
        if let retryAfter, retryAfter > delay { delay = retryAfter }
        nextDue = clock.now.addingTimeInterval(delay)
    }
}
